### PR TITLE
feat: context-aware entity selection for discovery prompts

### LINKED ISSUE #233: feat: context-aware entity selection for discovery prompts (location, relationships, co-occurrence)
## Summary

Replace recency-only entity context selection with context-aware selection that uses the narrative's current state (who, where, what) to choose which entities go into the discovery prompt's known-entity list.

## Motivation

Issue #221 introduces a token budget for the known-entity list, preventing unbounded context growth. The initial implementation uses recency (last_updated_turn) to prioritize which entities fill the budget. This works but has a known blind spot: entities that are semantically relevant to the current turn but haven't been mentioned recently get pruned.

Example: The player enters the Quiet Weave settlement at turn-280. The people, buildings, and historical facts associated with that location haven't been mentioned since turn-150. Recency-based pruning drops them, but they are exactly the entities needed for accurate coreference and discovery in this turn.

## Proposed Approach

Use the existing catalog's structured data (locations, relationships, events) to select contextually relevant entities, not just recent ones. No new infrastructure (no embedding model, no vector store) is needed — the catalog already contains the graph.

### Selection priority within the entity context budget:

1. **Mentioned** — entities whose names appear in the current turn text (keyword match)
2. **Co-located** — for mentioned locations, include their related characters, items, factions
3. **One-hop relationships** — for mentioned characters, include their relationship targets
4. **Location peers** — for included entities with a known location, include other entities at that location
5. **Recency backfill** — fill remaining budget with the most recently updated entities not yet included

### Data sources (all already exist in catalogs):

- `volatile_state.location` on character entities
- `relationships[].target_id` for one-hop graph traversal
- `related_entities` arrays in events for co-occurrence proximity
- Entity names for keyword matching against turn text

### Budget enforcement:

The token budget mechanism from #221 still applies. Context-aware selection decides *what* fills the budget; the budget cap prevents graph expansion from overflowing the context window.

### Graph expansion guard:

Dense locations or highly-connected entities could pull in 100+ entities through traversal. Priority ordering within the budget prevents this — mentioned entities always win, then co-located, then one-hop, then recency. If the budget fills at tier 2, tier 3-5 entities are simply omitted.

## Why not embedding-based retrieval?

The design study (Section 7, Investigation #1) suggests retrieval-augmented context as the most promising approach. However, for this domain:

- Fantasy names have poor embedding coverage
- Relationship links capture semantic connections that embeddings miss ("Kael's apprentice" won't embed near "Kael" but the catalog link is explicit)
- Location co-presence is structural, not textual
- No new dependencies required

Embedding-based retrieval remains a future option if graph-based selection proves insufficient.

## Affected Code

- `tools/catalog_merger.py` — `format_known_entities()` (or its replacement from #221)
- `tools/semantic_extraction.py` — entity discovery call site, where turn text and catalogs are both available

## Depends On

- #221 — bounded entity context (provides the budget mechanism this builds on)

## Acceptance Criteria

- [ ] Entity context selection considers current turn's mentioned entities, locations, and relationships
- [ ] Co-located and one-hop-related entities are pulled into context even if not recently updated
- [ ] Budget cap is respected regardless of graph traversal depth
- [ ] Entity context remains well-formed for the discovery template
- [ ] Unit tests verify priority ordering and budget enforcement
- [ ] No new external dependencies required

### PR DESCRIPTION
**Pull request by daviburg**
## Summary

Replace recency-only entity context selection with context-aware selection that uses the current turn's text to choose which entities fill the discovery prompt's known-entity list (#233).

### Problem

The bounded entity context mechanism (#221) uses recency to prioritize entities within the token budget. When the party revisits locations from earlier turns, the entities associated with those locations are pruned by recency while narratively irrelevant but recently-mentioned entities fill the budget. Field measurements on B70 showed 200+ irrelevant entities in discovery prompts while location-relevant ones were dropped.

### Solution

Multi-tier context-aware selection in `format_known_entities_bounded()`:

1. **Mentioned** — entities whose name or alias appears in the turn text (case-insensitive substring match; names < 3 chars skipped to avoid false positives)
2. **Co-located** — entities sharing a `volatile_state.location` with a mentioned entity, or entities at a mentioned location
3. **One-hop relationships** — `relationships[].target_id` targets of mentioned entities
4. **Recency backfill** — remaining entities sorted by `last_updated_turn` descending

Context-relevant entities (tiers 1-3) receive full detail regardless of recency window. Budget enforcement from #221 is preserved unchanged. When `turn_text` is not provided (e.g. legacy callers), falls back to the original recency-only behavior.

### Changes

- `tools/catalog_merger.py` — Added `_find_mentioned_entities()`, `_entity_names()`, `_get_entity_location()`, `_select_context_aware_entities()` helpers; extended `format_known_entities_bounded()` with `turn_text` parameter and context-aware selection logic
- `tools/semantic_extraction.py` — Pass `turn.get("text")` to `format_known_entities_bounded()` at the discovery call site
- `tests/test_context_aware_entity_selection.py` — 29 new tests covering priority ordering, budget enforcement, co-location, one-hop relationships, edge cases
- `docs/architecture.md` — Document context-aware entity selection under Semantic Extraction Layer
- `docs/roadmap.md` — Add entry under extraction quality passes

### Test Results

1441 passed (1412 baseline + 29 new), 0 regressions.

Closes #233

### CODE REVIEW

File: tests/test_context_aware_entity_selection.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+                identity="A long description " * 10,
+                last_updated_turn=f"turn-{i:03d}"))
+        catalogs = _make_catalogs(entities)
+        turn_text = " ".join(f"Entity{i:03d}" for i in range(50))
+        result = format_known_entities_bounded(
+            catalogs, current_turn=100,
+            entity_context_budget=200,
+            recency_window=10,
+            turn_text=turn_text)
+        tokens = _estimate_tokens(result)
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_context_aware_entity_selection.py]: Fixed in bf37f58 — removed the unused `tokens` variable.

File: tools/catalog_merger.py [hunk 2]
```diff
@@ -454,59 +617,95 @@
...
-    for entity in dormant:
-        line = _format_entity_brief(entity)
-        line_cost = _estimate_tokens(line + "\n")
-        if used + line_cost <= budget:
-            lines.append(line)
-            used += line_cost
-        else:
+    if used > budget and len(lines) > 1:
+        while lines and used > budget:
+            removed = lines.pop()
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — removed the unused `removed` variable.

File: tools/catalog_merger.py [hunk 3]
```diff
@@ -393,6 +393,158 @@ def _format_entity_brief(entity: dict) -> str:
...
+    if not turn_text:
+        return set()
+
+    text_lower = turn_text.lower()
+    mentioned: set[str] = set()
+    for entity in all_entities:
+        for name in _entity_names(entity):
+            if len(name) < _MIN_NAME_LENGTH_FOR_MATCH:
+                continue
+            if name.lower() in text_lower:
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_find_mentioned_entities()` now uses word-boundary-aware regex (`\b` for single-word names, `(?<!\w)…(?!\w)` for multi-word names). Names shorter than `_MIN_NAME_LENGTH_FOR_MATCH` (3 chars) are still skipped. Residual false positives from common English words used as entity names are tracked in #297.

File: tools/catalog_merger.py [hunk 4]
```diff
@@ -393,6 +393,158 @@ def _format_entity_brief(entity: dict) -> str:
...
+            entity = by_id.get(eid)
+            if entity is None:
+                continue
+            # If a mentioned entity IS a location, its name/aliases are
+            # location references
+            if entity.get("type") == "location":
+                for n in _entity_names(entity):
+                    relevant_locations.add(n.lower())
+            # If a mentioned entity has a location, that location is relevant
+            loc = _get_entity_location(entity)
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_collect_relevant_locations()` now includes both entity names/aliases AND `loc-*` entity IDs in the relevant locations set, plus the `volatile_state.location` value of every mentioned entity. This ensures matching works whether catalogs store locations as human-readable names or as `loc-*` IDs.

File: tools/catalog_merger.py [hunk 5]
```diff
@@ -393,6 +393,158 @@ def _format_entity_brief(entity: dict) -> str:
...
+    if mentioned_ids:
+        for eid in mentioned_ids:
+            entity = by_id.get(eid)
+            if entity is None:
+                continue
+            for rel in entity.get("relationships") or []:
+                target = rel.get("target_id")
+                if (target and target in by_id
+                        and target not in mentioned_ids
+                        and target not in colocated_ids):
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_find_one_hop_targets()` now filters on `status == "active"` (or unset). Dormant and resolved relationships are skipped.

File: tools/catalog_merger.py [hunk 6]
```diff
@@ -454,59 +617,95 @@ def format_known_entities_bounded(
...
+                relevant_locations.add(loc.lower())
+        for ent in all_entities:
+            if ent["id"] in mentioned_ids:
+                continue
+            loc = _get_entity_location(ent)
+            if loc and loc.lower() in relevant_locations:
+                context_ids.add(ent["id"])
+            if ent.get("type") == "location":
+                for n in _entity_names(ent):
+                    if n.lower() in relevant_locations:
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — the duplicate location recomputation in `format_known_entities_bounded()` was eliminated. The function now uses the `priority_ids` set returned by `_select_context_aware_entities()` directly, so the fix in `_collect_relevant_locations()` applies consistently.

File: tools/catalog_merger.py [hunk 7]
```diff
@@ -454,59 +617,95 @@ def format_known_entities_bounded(
...
+                        break
+        # One-hop
+        for eid in mentioned_ids:
+            ent = by_id.get(eid)
+            if not ent:
+                continue
+            for rel in ent.get("relationships") or []:
+                target = rel.get("target_id")
+                if target and target in by_id and target not in mentioned_ids:
+                    context_ids.add(target)
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — same as above. The duplicate one-hop pass was removed; priority is determined once in `_select_context_aware_entities()` which already uses the active-only filter.

File: tests/test_context_aware_entity_selection.py [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+    format_known_entities_bounded,
+    _estimate_tokens,
+    _find_mentioned_entities,
+    _entity_names,
+    _get_entity_location,
+    _select_context_aware_entities,
+    _collect_relevant_locations,
+    _find_colocated_entities,
+    _find_one_hop_targets,
+)
```
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_context_aware_entity_selection.py:14]: Fixed in b9c31df — removed the unused imports (`_collect_relevant_locations`, `_find_colocated_entities`, `_find_one_hop_targets`).

File: tests/test_context_aware_entity_selection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_context_aware_entity_selection.py]: Fixed in bf37f58 — removed the unused `tokens` variable.

File: tools/catalog_merger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — removed the unused `removed` variable.

File: tools/catalog_merger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_find_mentioned_entities()` now uses word-boundary-aware regex (`\b` for single-word names, `(?<!\w)…(?!\w)` for multi-word names). Names shorter than `_MIN_NAME_LENGTH_FOR_MATCH` (3 chars) are still skipped. Residual false positives from common English words used as entity names are tracked in #297.

File: tools/catalog_merger.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_collect_relevant_locations()` now includes both entity names/aliases AND `loc-*` entity IDs in the relevant locations set, plus the `volatile_state.location` value of every mentioned entity. This ensures matching works whether catalogs store locations as human-readable names or as `loc-*` IDs.

File: tools/catalog_merger.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_find_one_hop_targets()` now filters on `status == "active"` (or unset). Dormant and resolved relationships are skipped.

File: tools/catalog_merger.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — the duplicate location recomputation in `format_known_entities_bounded()` was eliminated. The function now uses the `priority_ids` set returned by `_select_context_aware_entities()` directly, so the fix in `_collect_relevant_locations()` applies consistently.

File: tools/catalog_merger.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — same as above. The duplicate one-hop pass was removed; priority is determined once in `_select_context_aware_entities()` which already uses the active-only filter.

File: tests/test_context_aware_entity_selection.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_context_aware_entity_selection.py:14]: Fixed in b9c31df — removed the unused imports (`_collect_relevant_locations`, `_find_colocated_entities`, `_find_one_hop_targets`).

File: tests/test_context_aware_entity_selection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_context_aware_entity_selection.py]: Fixed in bf37f58 — removed the unused `tokens` variable.

File: tools/catalog_merger.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — removed the unused `removed` variable.

File: tools/catalog_merger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_find_mentioned_entities()` now uses word-boundary-aware regex (`\b` for single-word names, `(?<!\w)…(?!\w)` for multi-word names). Names shorter than `_MIN_NAME_LENGTH_FOR_MATCH` (3 chars) are still skipped. Residual false positives from common English words used as entity names are tracked in #297.

File: tools/catalog_merger.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_collect_relevant_locations()` now includes both entity names/aliases AND `loc-*` entity IDs in the relevant locations set, plus the `volatile_state.location` value of every mentioned entity. This ensures matching works whether catalogs store locations as human-readable names or as `loc-*` IDs.

File: tools/catalog_merger.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — `_find_one_hop_targets()` now filters on `status == "active"` (or unset). Dormant and resolved relationships are skipped.

File: tools/catalog_merger.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — the duplicate location recomputation in `format_known_entities_bounded()` was eliminated. The function now uses the `priority_ids` set returned by `_select_context_aware_entities()` directly, so the fix in `_collect_relevant_locations()` applies consistently.

File: tools/catalog_merger.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tools/catalog_merger.py]: Fixed in bf37f58 — same as above. The duplicate one-hop pass was removed; priority is determined once in `_select_context_aware_entities()` which already uses the active-only filter.

File: tests/test_context_aware_entity_selection.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply tests/test_context_aware_entity_selection.py:14]: Fixed in b9c31df — removed the unused imports (`_collect_relevant_locations`, `_find_colocated_entities`, `_find_one_hop_targets`).
